### PR TITLE
Kernel: x86_64 support (part 4)

### PR DESCRIPTION
**Pull request by gunnarbeutner**
**Kernel: Fix incorrect flags for the GDT entries**

The Sz (protected mode) bit should not be set for 64-bit GDT entries.

**Kernel: Fix declaration for struct TrapFrame**

This is declared as a struct elsewhere.

**Kernel+Userland: Add x86_64 registers to RegisterState/PtraceRegisters**

**Kernel: Add support for interrupts on x86_64**

**Kernel: Use macros for GDT indices instead of hardcoding the values**

**Kernel: Clear segment registers on x86_64**

They're supposedly unused but iretq doesn't like ss being non-zero.

**Kernel: Implement safe_\* memory access functions for x86_64**

**Kernel: Reorder code a bit to clarify which #if block it belongs to**

**Kernel: Rename Thread::tss to Thread::regs and add x86_64 support**

We're using software context switches so calling this struct tss is somewhat misleading.

**Kernel: Implement initializing threads on x86_64**

### CODE REVIEW

File: Kernel/Arch/x86/TrapFrame.h [hunk 1]
```diff
@@ -15,7 +15,11 @@
 namespace Kernel {
 
 struct TrapFrame {
+#if ARCH(I386)
     u32 prev_irq_level;
+#else
+    u64 prev_irq_level;
+#endif
```
Comment (Hendiadyoin1) [inline Kernel/Arch/x86/TrapFrame.h:22]: FlatPtr?

Comment (gunnarbeutner) [reply Kernel/Arch/x86/TrapFrame.h]: Fixed. :)

File: Kernel/Arch/x86/x86_64/Processor.cpp [hunk 2]
```diff
@@ -42,6 +41,7 @@ asm(
 asm(
 ".global do_assume_context \n"
 "do_assume_context: \n"
+"    int 3 \n"
```
Comment (IdanHo) [inline Kernel/Arch/x86/x86_64/Processor.cpp:44]: is this meant to be here? (amazing work btw :)

Comment (gunnarbeutner) [reply Kernel/Arch/x86/x86_64/Processor.cpp]: Thank you! Also, no, but it doesn't work atm either way. I've removed it for now. :)

File: Kernel/Arch/x86/common/Processor.cpp [hunk 3]
```diff
@@ -1067,8 +1067,8 @@ UNMAP_AFTER_INIT void Processor::gdt_init()
     write_raw_gdt_entry(GDT_SELECTOR_CODE3, 0x0000ffff, 0x00cffa00); // code3
     write_raw_gdt_entry(GDT_SELECTOR_DATA3, 0x0000ffff, 0x00cff200); // data3
 #else
-    write_raw_gdt_entry(GDT_SELECTOR_CODE0, 0x0000ffff, 0x00ef9a00); // code0
-    write_raw_gdt_entry(GDT_SELECTOR_CODE3, 0x0000ffff, 0x00effa00); // code3
+    write_raw_gdt_entry(GDT_SELECTOR_CODE0, 0x0000ffff, 0x00af9a00); // code0
+    write_raw_gdt_entry(GDT_SELECTOR_CODE3, 0x0000ffff, 0x00affa00); // code3
```
Comment (Hendiadyoin1) [inline Kernel/Arch/x86/common/Processor.cpp:1070]: maybe we should get away form initializing them raw, and initialize them with proper arguments instead

Just a nit

**Pull request merged**